FlowCrypt/Core/Core.swift

```diff
@@ -236,47 +230,29 @@ final class Core: KeyDecrypter, KeyParser, CoreComposeMessageType {
     }
 
     private func call(_ endpoint: String, jsonData: Data, data: Data) async throws -> RawRes {
-        try await sleepUntilReadyOrThrow()
-        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<RawRes, Error>) in
-            // tom - todo - currently there is only one callback storage variable "cb_last_value"
-            //   for all JavaScript calls, and so we have to synchronize
-            //   all calls into JavaScript to happen serially, else
-            //   the return values would be undefined when used concurrently
-            //   see https://github.com/FlowCrypt/flowcrypt-ios/issues/852
-            // A possible solution would be to only synchronize returning o fthe callbac values into some dict. But I'm unsure if JavaScript is otherwise safe to call concurrently, so for now we'll do the safer thing.
-            queue.async {
-                self.cb_last_value = nil
-                self.jsEndpointListener!.call(withArguments: [endpoint, String(data: jsonData, encoding: .utf8)!, Array<UInt8>(data), self.cb_catcher!])
-                guard
-                    let resJsonData = self.cb_last_value?.0.data(using: .utf8),
-                    let rawResponse = self.cb_last_value?.1
-                else {
-                    self.logger.logError("could not see callback response, got cb_last_value: \(String(describing: self.cb_last_value))")
-                    continuation.resume(throwing: CoreError.format("JavaScript callback response not available"))
-                    return
-                }
-                let error = try? resJsonData.decodeJson(as: CoreRes.Error.self)
-                if let error = error {
-                    let errMsg = "------ js err -------\nCore \(endpoint):\n\(error.error.message)\n\(error.error.stack ?? "no stack")\n------- end js err -----"
-                    self.logger.logError(errMsg)
-                    continuation.resume(throwing: CoreError(coreError: error))
-                    return
-                }
-                continuation.resume(returning: RawRes(json: resJsonData, data: Data(rawResponse)))
-            }
-        }
-    }
-    
-    private func sleepUntilReadyOrThrow() async throws {
-        // This will block the task for up to 1000ms if the app was just started and Core method was called before JSContext is ready
-        // It should only affect the user if Core method was called within 500-800ms of starting the app
-        let start = DispatchTime.now()
-        while !ready {
-            if start.millisecondsSince > 1000 { // already waited for 1000 ms, give up
-                throw CoreError.notReady("App Core not ready yet")
-            }
+        jsEndpointListener!.call(withArguments: [endpoint, String(data: jsonData, encoding: .utf8)!, Array<UInt8>(data), cb_catcher!])
+
+        while callbackResults.isEmpty {
```

Comment:
Since we have `CoreHost` to allow JavaScript world communicate with us, we have to call `handleCallbackResult()` method inside task. This leads to empty callback result because that task will be executed after we leave this method. To allow that task to be executed we need to "sleep" here for some small time interval.

Reply:
It doesn't feel right - can you reproduce this situation with empty callback also before your refactor? I've never seen that happen.

Reply:
It is possible if you run several `Core` methods in parallel. For example this was the reason why `testCoreResponseCorrectnessUnderConcurrency()` test failed on my machine

Reply:
So without this delay we can not get any value from call-back, because while we are inside this `Core` method, any call to other `Core` methods will wait. To give `Core.handleCallbackResult()` method a chance for invocation we `await` on timer.

Reply:
But isn't the whole point of `actor` that only one async method can happen in parallel?

Or are you saying that several methods can work in parallel, but they are synchronized on the same thread?

Reply:
I see - answer is yes - I got your answer below